### ISSUE TITLE
pixracer mag offset

### ISSUE THREAD

**Issue by urlu75** [PX4/Firmware#7358]
I was using PX4 1.5.5 without any major issue (hexacopter). Upgrade to 1.6.x and now have an 180 degrees offset with the mag after recalibrating all the sensors.
I don't know if it's related, but I can't go down when altitude hold is enable and position control is flying away (probably the mag issue).

Downgrade to 1.5.5 with the same setting, recalibrate all sensors and everything is fine again.

**Comment by mhkabir** [comment]
Are you running 1.6.1 which we released yesterday? We fixed an issue with the mag rotation there. 

So please describe your setup, external mag rotation, etc.

**Comment by urlu75** [comment]
I tried the 1.6.1 this afternoon.
I do not have the hexa with me to dump the setting (sorry for the low detail).
The board is front/back inverted (yaw180) and external compass (HMC5983) has no rotation (YAW 0). X4R SBUS receiver, frsky telemetry on frsky port, mavlink telemetry for OSD on port 1 and normal telemetry @19k on port 2. Almost all the parameters were by default with this testing.

**Comment by mhkabir** [comment]
@urlu75 Did you reboot after calibration? Please provide a parameter dump too.

**Comment by matanhavi** [comment]
@mhkabir note that there still a bug.
if you change parameter (any parameter, for example I changed the SDLOG_UTC_OFFSET) before flight, for some reason the SENS_BOARD_ROT 'loses' its value and the rotation is wrong.
you can easily test it on table, put SENS_BOARD_ROT as non default value (i had 2), reset,
look at the heading, calibrate if need.
change SDLOG_UTC_OFFSET, and look at the heading.

**Comment by mhkabir** [comment]
@matanhavi OK that parameter not saving thing sounds concerning. Let me have a look.

**Comment by mhkabir** [comment]
I just checked both your issues - 

> if you change parameter (any parameter, for example I changed the SDLOG_UTC_OFFSET) before flight, for some reason the SENS_BOARD_ROT 'loses' its value and the rotation goes wrong.
you can easily test it on table, put SENS_BOARD_ROT as non default value (i had 2), reset,
look at the heading, calibrate if need.
change SDLOG_UTC_OFFSET, and look at the heading.

@matanhavi I could not reproduce this. What hardware are you running? From what vendor?

> Upgrade to 1.6.x and now have an 180 degrees offset with the mag after recalibrating all the sensors.

@urlu75 I just flashed a fresh board and replicated your sensor setup. Everything looks fine on my side and I wasn't able to reproduce it - I extensively tested a bunch of rotations and corner cases. Please reflash your board using QGC with 1.6.1 to make sure you're on the latest version.

**Comment by matanhavi** [comment]
@mhkabir I am using regular Pixhawk 2.1. and version 1.6.1
Try reproduce by this:
- use the Pixhawk on table, no other HW needed (also not external compass)
- set SENS_BOARD_ROT to YAW 90
- calibrate (if needed)
- reset
- look at highres_IMU.xmag,ymag,zmag
- change SDLOG_UTC_OFFSET to anything
- look at highres_IMU.xmag,ymag,zmag, ymag goes to xmag value, xmag goes to -xmag
![image](https://cloud.githubusercontent.com/assets/1760878/26819740/e9fe040a-4aa8-11e7-95ad-45696b2887fe.png)
- if you reset again, all goes back to "normal"

**Comment by mhkabir** [comment]
Still couldn't reproduce. I tried with a Pixhawk 2.1 : 

![image](https://cloud.githubusercontent.com/assets/5557676/26820294/ff0572b8-4abf-11e7-9383-f03ffd237061.png)

**Comment by matanhavi** [comment]
I done some more test, from fresh parameters of Pixhawk i couldn't reproduce it too.
but after I set my parameter file, it does reproduced. I am trying to find why

**Comment by matanhavi** [comment]
I handpick some parameters from my AP with that issue.
and reproduce it with minimal changes from default parameters of new AP. (airframe not even set... and calibration is not real, i toke it from other AP)
I attached both files, on that is just default parameters after set to default from QGC
the other is one that i changed and reproduce the issue
[P_badparams.txt](https://github.com/PX4/Firmware/files/1054674/P_badparams.txt)
[default_parmas.txt](https://github.com/PX4/Firmware/files/1054675/default_parmas.txt)

**Comment by matanhavi** [comment]
@mhkabir I have more info.
its seems that at first when the AP load it uses SENS_BOARD_ROT as the internal magnetometer rotation.
but after some change of parameter, VotedSensorsUpdate::parameters_update() is called, but there, the CAL_MAG0_ROT is used instead.

i think it might be related to on of your last commits, 
https://github.com/PX4/Firmware/commit/6ee09ca16fd5e52237b4566f6d10de525cc2cb9e

**Comment by mhkabir** [comment]
The thing is - after a change of parameters, you are supposed to reboot. QGC doesn't force this yet, but we will add it soon. Can you see if you can reproduce the issue if you reboot the autopilot after any changes to a the CAL_MAG* or SENS_BOARD_ROT parameter(s)?

**Comment by mhkabir** [comment]
@mhkabir OK I think I found a bug. Stay tuned for testing the fix.

**Comment by matanhavi** [comment]
"The thing is - after a change of parameters, you are supposed to reboot."  - That is not true for few reasons:
1) there are some internal logics that changes parameters, for example, the flight time, BAT_N_CELLS, and other (there is even param_set_no_notification() on this code)
2) how do you supposed to calibrate PID, for example with this restriction
3) the reason for each 'parameters_update()' on modules is to allow parameter change on the fly, those logics must be safe.

I agree that for some parameters its advised not to change "on air" and you sure must to reboot. having mechanism to lock them can be nice.

I am not sure now about my previous comment.

**Comment by mhkabir** [comment]
What I meant was, after changing the calibration and orientation parameters - https://github.com/PX4/Firmware/commit/49890c61f5c98be9aaf5371fb5d29a8168b2a0d6

Please do as I say and try rebooting after setting orientations and/or calibrating and see if you can still reproduce the bad mag rotation problem.

**Comment by matanhavi** [comment]
I am rebooting after calibrate. calibrating goes to "I agree that for some parameters its advised not to change "on air" and you sure must to reboot" statement.
on all my tests I didn't calibrate at all.
Can you go to gitter now?

**Comment by mhkabir** [comment]
Sure. Let's chat.

**Comment by mhkabir** [comment]
OK I found the issue. Some old code in `parameters_update()` was resetting the rotation matrices unnecessarily. I fixed it here : https://github.com/PX4/Firmware/pull/7366

I and the tesflights team (when they flew the original fix) weren't initially able to reproduce it locally because it would only trigger depending on sensor startup timing (happens around 1/10 times). Sorry about all the drama regarding this fix :P

We still need QGC to force a reboot on setting mag and/or board orientation after calibration. @DonLakeFlyer @LorenzMeier
